File: src/lib/sources/openweather.ts
import {
  WeatherSource,
  CurrentData,
  HourlyData,
  DailyData,
  getConditionIcon,
  getDayNameRo,
} from "./base";

const BASE_URL = "https://api.openweathermap.org/data/2.5";
const TIMEOUT = 15000;

const OWM_MAP: Record<string, string> = {
  Clear: "senin",
  Clouds: "noros",
  Drizzle: "burinta_usoara",
  Rain: "ploaie_moderata",
  Thunderstorm: "furtuna",
  Snow: "ninsoare_moderata",
  Mist: "ceata",
  Fog: "ceata",
  Haze: "ceata",
};

function mapCondition(weather: Array<{ main: string }>): string {
  if (!weather?.[0]) return "necunoscut";
  return OWM_MAP[weather[0].main] || "necunoscut";
}

export class OpenWeatherSource implements WeatherSource {
  name = "openweather";

  private get apiKey(): string {
    return process.env.OPENWEATHER_API_KEY || "";
  }

  async fetchCurrent(lat: number, lon: number): Promise<CurrentData> {
    if (!this.apiKey) throw new Error("OpenWeather API key lipsă");

    const params = new URLSearchParams({
      lat: lat.toString(),
      lon: lon.toString(),
      units: "metric",
      appid: this.apiKey,
    });

    const resp = await fetch(`${BASE_URL}/weather?${params}`, {
      signal: AbortSignal.timeout(TIMEOUT),
    });
    const data = await resp.json();
    if (data.cod && data.cod !== 200) throw new Error(data.message);

    const condition = mapCondition(data.weather);
    return {
      temperature: data.main?.temp ?? 0,
      humidity: data.main?.humidity ?? 0,
      wind_speed: Math.round((data.wind?.speed ?? 0) * 3.6 * 10) / 10,
      wind_direction: data.wind?.deg ?? 0,
      precipitation: data.rain?.["1h"] ?? 0,
      condition,
      condition_icon: getConditionIcon(condition),
    };
  }

  async fetchHourly(lat: number, lon: number): Promise<HourlyData[]> {
    if (!this.apiKey) throw new Error("OpenWeather API key lipsă");

    const params = new URLSearchParams({
      lat: lat.toString(),
      lon: lon.toString(),
      units: "metric",
      appid: this.apiKey,
    });

    const resp = await fetch(`${BASE_URL}/forecast?${params}`, {
      signal: AbortSignal.timeout(TIMEOUT),
    });
    const data = await resp.json();

    return (data.list || [])
      .slice(0, 8)
      .map((item: Record<string, unknown>) => {
        const dt = new Date((item.dt as number) * 1000);
        const condition = mapCondition(
          (item as { weather: Array<{ main: string }> }).weather,
        );
        const main = item.main as { temp: number; humidity: number };
        const wind = item.wind as { speed: number };
        const rain = item.rain as Record<string, number> | undefined;
        return {
          hour: dt.toLocaleTimeString("ro-RO", {
            hour: "2-digit",
            minute: "2-digit",
            timeZone: "Europe/Bucharest",
          }),
          timestamp: dt.toISOString(),
          temperature: main?.temp ?? 0,
          humidity: main?.humidity ?? 0,
          wind_speed: Math.round((wind?.speed ?? 0) * 3.6 * 10) / 10,
          precipitation: rain?.["3h"] ?? 0,
          condition,
        };
      });
  }

  async fetchDaily(lat: number, lon: number): Promise<DailyData[]> {
    if (!this.apiKey) throw new Error("OpenWeather API key lipsă");

    const params = new URLSearchParams({
      lat: lat.toString(),
      lon: lon.toString(),
      units: "metric",
      appid: this.apiKey,
    });

    const resp = await fetch(`${BASE_URL}/forecast?${params}`, {
      signal: AbortSignal.timeout(TIMEOUT),
    });
    const data = await resp.json();

    const days: Record<string, Array<Record<string, unknown>>> = {};
    for (const item of data.list || []) {
      const dt = new Date((item.dt as number) * 1000);
      const key = dt.toISOString().split("T")[0];
      if (!days[key]) days[key] = [];
      days[key].push(item);
    }

    return Object.entries(days)
      .slice(0, 5)
      .map(([dateStr, items]) => {
        const temps = items.map((it) => (it.main as { temp: number }).temp);
        const midItem = items[Math.floor(items.length / 2)];
        const condition = mapCondition(
          (midItem as { weather: Array<{ main: string }> }).weather,
        );
        return {
          date: dateStr,
          day_name: getDayNameRo(dateStr),
          temp_min: Math.round(Math.min(...temps) * 10) / 10,
          temp_max: Math.round(Math.max(...temps) * 10) / 10,
          precipitation: items.reduce(
            (sum, it) =>
              sum +
              ((it.rain as Record<string, number> | undefined)?.["3h"] ?? 0),
            0,
          ),
          wind_speed:
            Math.round(
              Math.max(
                ...items.map(
                  (it) => ((it.wind as { speed: number })?.speed ?? 0) * 3.6,
                ),
              ) * 10,
            ) / 10,
          condition,
        };
      });
  }
}

File: backend/sources/openweather.py
import httpx
from .base import WeatherSource, get_condition_icon
from config import SOURCE_TIMEOUT, OPENWEATHER_API_KEY

# Mapare conditii OpenWeatherMap -> conditii normalizate
OWM_CONDITIONS = {
    "Clear": "senin",
    "Clouds": "noros",
    "Few clouds": "partial_noros",
    "Scattered clouds": "partial_noros",
    "Broken clouds": "noros",
    "Overcast clouds": "noros",
    "Drizzle": "burinta_usoara",
    "Rain": "ploaie_moderata",
    "Light rain": "ploaie_usoara",
    "Moderate rain": "ploaie_moderata",
    "Heavy rain": "ploaie_puternica",
    "Thunderstorm": "furtuna",
    "Snow": "ninsoare_moderata",
    "Light snow": "ninsoare_usoara",
    "Heavy snow": "ninsoare_puternica",
    "Mist": "ceata",
    "Fog": "ceata",
    "Haze": "ceata",
}


def _map_condition(weather_data: dict) -> str:
    if not weather_data.get("weather"):
        return "necunoscut"
    main = weather_data["weather"][0].get("main", "")
    desc = weather_data["weather"][0].get("description", "")
    return OWM_CONDITIONS.get(desc.capitalize(), OWM_CONDITIONS.get(main, "necunoscut"))


class OpenWeatherSource(WeatherSource):
    name = "openweather"
    BASE_URL = "https://api.openweathermap.org/data/2.5"

    async def fetch_current(self, lat: float, lon: float) -> dict:
        if not OPENWEATHER_API_KEY:
            raise ValueError("OpenWeatherMap API key nu este configurat")

        params = {
            "lat": lat,
            "lon": lon,
            "units": "metric",
            "appid": OPENWEATHER_API_KEY,
        }
        async with httpx.AsyncClient(timeout=SOURCE_TIMEOUT) as client:
            resp = await client.get(f"{self.BASE_URL}/weather", params=params)
            resp.raise_for_status()
            data = resp.json()

        condition = _map_condition(data)
        return {
            "temperature": data["main"]["temp"],
            "humidity": data["main"]["humidity"],
            "wind_speed": data["wind"]["speed"] * 3.6,  # m/s -> km/h
            "wind_direction": data["wind"].get("deg", 0),
            "precipitation": data.get("rain", {}).get("1h", 0),
            "condition": condition,
            "condition_icon": get_condition_icon(condition),
        }

    async def fetch_hourly(self, lat: float, lon: float) -> list[dict]:
        if not OPENWEATHER_API_KEY:
            raise ValueError("OpenWeatherMap API key nu este configurat")

        params = {
            "lat": lat,
            "lon": lon,
            "units": "metric",
            "appid": OPENWEATHER_API_KEY,
        }
        async with httpx.AsyncClient(timeout=SOURCE_TIMEOUT) as client:
            resp = await client.get(f"{self.BASE_URL}/forecast", params=params)
            resp.raise_for_status()
            data = resp.json()

        from datetime import datetime
        import pytz
        tz = pytz.timezone("Europe/Bucharest")
        results = []
        for item in data.get("list", [])[:8]:  # Primele 24h (8 x 3h)
            dt = datetime.fromtimestamp(item["dt"], tz=tz)
            condition = _map_condition(item)
            results.append({
                "hour": dt.strftime("%H:%M"),
                "timestamp": dt.isoformat(),
                "temperature": item["main"]["temp"],
                "humidity": item["main"]["humidity"],
                "wind_speed": item["wind"]["speed"] * 3.6,
                "precipitation": item.get("rain", {}).get("3h", 0),
                "condition": condition,
            })
        return results

    async def fetch_daily(self, lat: float, lon: float) -> list[dict]:
        if not OPENWEATHER_API_KEY:
            raise ValueError("OpenWeatherMap API key nu este configurat")

        params = {
            "lat": lat,
            "lon": lon,
            "units": "metric",
            "appid": OPENWEATHER_API_KEY,
        }
        async with httpx.AsyncClient(timeout=SOURCE_TIMEOUT) as client:
            resp = await client.get(f"{self.BASE_URL}/forecast", params=params)
            resp.raise_for_status()
            data = resp.json()

        from datetime import datetime
        import pytz
        tz = pytz.timezone("Europe/Bucharest")
        day_names_ro = ["Luni", "Marți", "Miercuri", "Joi", "Vineri", "Sâmbătă", "Duminică"]

        # Grupeaza pe zile
        days: dict[str, list] = {}
        for item in data.get("list", []):
            dt = datetime.fromtimestamp(item["dt"], tz=tz)
            date_key = dt.strftime("%Y-%m-%d")
            if date_key not in days:
                days[date_key] = []
            days[date_key].append(item)

        results = []
        for date_str, items in list(days.items())[:5]:
            temps = [it["main"]["temp"] for it in items]
            dt = datetime.strptime(date_str, "%Y-%m-%d")
            day_name = day_names_ro[dt.weekday()]
            condition = _map_condition(items[len(items) // 2])
            results.append({
                "date": date_str,
                "day_name": day_name,
                "temp_min": round(min(temps), 1),
                "temp_max": round(max(temps), 1),
                "precipitation": sum(it.get("rain", {}).get("3h", 0) for it in items),
                "wind_speed": max(it["wind"]["speed"] * 3.6 for it in items),
                "condition": condition,
            })
        return results
